src/cuda/backend.f90

```diff
@@ -228,42 +232,45 @@ subroutine transeq_cuda_dist(self, du, dv, dw, u, v, w, dirps, &
     type is (cuda_tdsops_t); der2nd_sym => tdsops
     end select
 
-    call transeq_halo_exchange(self, u_dev, v_dev, w_dev, dirps)
+    call transeq_halo_exchange(self, u_dev, v_dev, w_dev, dirps%dir)
 
     call transeq_dist_component(self, du_dev, u_dev, u_dev, &
                                 self%u_recv_s_dev, self%u_recv_e_dev, &
                                 self%u_recv_s_dev, self%u_recv_e_dev, &
-                                der1st, der1st_sym, der2nd, dirps, &
+                                der1st, der1st_sym, der2nd, dirps%dir, &
                                 blocks, threads)
     call transeq_dist_component(self, dv_dev, v_dev, u_dev, &
                                 self%v_recv_s_dev, self%v_recv_e_dev, &
                                 self%u_recv_s_dev, self%u_recv_e_dev, &
-                                der1st_sym, der1st, der2nd_sym, dirps, &
+                                der1st_sym, der1st, der2nd_sym, dirps%dir, &
                                 blocks, threads)
     call transeq_dist_component(self, dw_dev, w_dev, u_dev, &
                                 self%w_recv_s_dev, self%w_recv_e_dev, &
                                 self%u_recv_s_dev, self%u_recv_e_dev, &
-                                der1st_sym, der1st, der2nd_sym, dirps, &
+                                der1st_sym, der1st, der2nd_sym, dirps%dir, &
                                 blocks, threads)
 
   end subroutine transeq_cuda_dist
 
-  subroutine transeq_halo_exchange(self, u_dev, v_dev, w_dev, dirps)
+  subroutine transeq_halo_exchange(self, u_dev, v_dev, w_dev, dir)
     class(cuda_backend_t) :: self
     real(dp), device, dimension(:, :, :), intent(in) :: u_dev, v_dev, w_dev
-    type(dirps_t), intent(in) :: dirps
-    integer :: n_halo
+    integer, intent(in) :: dir
+    integer :: n_halo, n, nproc_dir, pprev, pnext
+    integer :: n_groups
 
     ! TODO: don't hardcode n_halo
     n_halo = 4
+    n_groups = self%mesh%get_n_groups(dir)
+    n = self%mesh%get_n(dir, VERT)
```

Comment:
I had to change `get_n(u)` to `get_n(dir, VERT)` here because in the CUDA backend at this level we don't have a field `u` available, and it didn't compile. This should work all fine, but later on we can try to find a better solution.